### ISSUE TITLE
2.2.21 broke initialMessages

### ISSUE THREAD

**Issue by odysseaspapadimas** [vercel/ai#724]
### Description

Updating from 2.2.20 to 2.2.21 broke the initialMessages property in the react useChat hook. The messages do not load.
As you can see from my code example I load the initial chat messages from my API which I checked and it returns the data just fine. Downgrading to 2.2.20 fixes it

### Code example

```jsx
const {
    messages,
    handleInputChange,
    isLoading,
    append,
  } = useChat({
    id: chatId,
    initialMessages: chat.data?.messages.map((message) => ({
      content: message.content,
      role: MessageRole(message.role),
      id: message.id,
    })),
});
```

**Comment by lgrammel** [comment]
Since `v2.2.21`, `initialMessages` behaves similar to `initialInput` and uses `useState` https://github.com/vercel/ai/commit/0681baeac2027188f9bac3718ad64e7d194887f0 to address issue #615 

If you initialize it with `undefined` first (e.g. if `chat.data` is not yet available), then you cannot override it later. Changing the initial messages can also lead to strange behaviors once a chat has started and there are actual messages.

That being said, there needs to be a way to deal with `initialMessage` that are not immediately available.

@odysseaspapadimas Just to confirm, is `chat.data?.messages` available on the first invocation of `useChat` or does it change (e.g. by calling an API like you mentioned)?

**Comment by odysseaspapadimas** [comment]
@lgrammel No, chat.data?.messages is undefined.

And yes I agree there needs to be a way to deal with this. Are there any solutions in the meantime other than downgrading?

**Comment by mikesprague** [comment]
@lgrammel in my case, it (`initialMessages`) starts as an empty array and gets updated based on a value in local storage (but is never actually `undefined`). The value changes value from `[]` to an array with the saved messages data.

**Comment by lgrammel** [comment]
@odysseaspapadimas @mikesprague A workaround for the time being would be to extract the component that contains `useChat` and only initalize it when `initialMessages` has been loaded (so it's correct the first time you call `useChat`)

**Comment by greendesertsnow** [comment]
I was going to lose my mind after updating the packages...
Wrap the whole thing in a messageLoader component and pass the props down only if you've got the old messages from your database.
Example (edited it, so it may not work):
```js
export default function Chat(props: any) {
  const { selectedChat } = props;
  const { data: messagesFromDb } = useMessages(selectedChat);

  return (
    <div>
      {messagesFromDb && <MessageLoader messagesFromDb={messagesFromDb} />}
    </div>
  );
}

export function MessageLoader(props: any) {
  const { messagesFromDb } = props;

  const {
    messages,
    input,
    reload,
    stop,
    isLoading,
    handleInputChange,
    handleSubmit,
    data: inputData,
  } = useChat({
    initialMessages: messagesFromDb,
  });

 
  const handleInputSubmit = (e: any) => {
    e.preventDefault();
    console.log({ isLoading });
    if (!isLoading) handleSubmit(e);
  };

  return (
    <div className="flex flex-col  max-w-2xl  pb-32 mx-auto stretch">
      <div className="">
        {messages.length > 0
          ? messages.map(m => (
              <div
                key={m.id}
                className={`whitespace-pre-wrap my-4 flex ${
                  m.role === 'user' ? 'justify-end' : null
                }`}
              >
             
                  <ChatMessage
                    user={user}
                    role={m.role}
                    content={m.content}
                    createdAt={m.createdAt}
                  />
               
              </div>
            ))
          : null}
      </div>
      <form onSubmit={handleInputSubmit} >
        <Textarea />
      </form>
    </div>
  );
}


export function ChatMessage({
  user,
  role,
  content,
  createdAt,
}: ChatMessageProps) {
  return ( <div>Chatmessages </div>)
}

```

**Comment by mosnicholas** [comment]
This should work similar to how useEffect or other hooks work, so that we can pass in an array of dependencies and it rerenders / recreates the initial messages if it changes. This is a pretty big breaking change as most people will likely be fetching messages from an API call and then loading them in to `useChat` call.

**Comment by MaxLeiter** [comment]
This should be addressed in 2.2.22. Thanks for the reports!

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This is a concrete regression in existing `useChat` behavior introduced by version 2.2.21, with a workaround and planned patch confirmed by maintainers.

### Evidence
- Updating from 2.2.20 to 2.2.21 caused `useChat` to retain the first `initialMessages` value, so asynchronously loaded messages were not displayed.
- Multiple reporters confirmed regressions when `initialMessages` changed from `undefined` or `[]` to fetched or locally stored messages.
- A maintainer acknowledged the need to support messages that are not immediately available and stated the issue should be addressed in 2.2.22.
